CaloMC/src/CaloNoiseSimGenerator.cc

```diff
@@ -0,0 +1,220 @@
+       if (enableAR_) initAR();
+       generateFragments(wfExtractor);
+   }
+
+   //------------------------------------------------------------------------------------------------------------------
+   void CaloNoiseSimGenerator::generateWF(std::vector<double>& wfVector)
+   {     
+       ConditionsHandle<CalorimeterCalibrations> calorimeterCalibrations("ignored");
+       double scaleFactor = calorimeterCalibrations->MeV2ADC(iRO_)/calorimeterCalibrations->peMeV(iRO_);
+
+       std::fill(wfVector.begin(),wfVector.end(),0);
+
+       const auto&      pulse         = pulseShape_.digitizedPulse(0.0);
+       const unsigned   pulseSize     = pulse.size();
+       const unsigned   bufferSize    = int(0.75*pulseSize);
+       const unsigned   noiseSize     = wfVector.size();
+       const double     totalTime     = (noiseSize+bufferSize)*digiSampling_;
+       const int        noiseLevelPE  = int(totalTime*noiseRinDark_);
+
+       //Generate the radiation induced noise (RIN)
+       const int nPh = randPoisson_(noiseLevelPE);
+       for (int i=0;i<nPh;++i)
+       {
+	   double t0 = randFlat_.fire(0.0,totalTime);       
+	   const auto& wf = pulseShape_.digitizedPulse(t0);
+
+	   int i0 = int(t0/digiSampling_) - bufferSize;
+	   int l0 = (i0<0) ? -i0 : 0;      
+	   int l1 = std::min(pulseSize,noiseSize-i0);
+	   for (int l=l0;l<l1;++l) wfVector[i0+l] += wf[l]*scaleFactor;
+       }
+
+       //add electronics noise        
+       double noiseADC = noiseElec_*digiSampling_*scaleFactor;
+       for (auto& val : wfVector) val += randGauss_.fire(0.0,noiseADC);
+
+       //estimate pedestal for this waveform - set it to theoretical value for the time being
+       pedestal_ = int(noiseRinDark_*digiSampling_*std::accumulate(pulse.begin(),pulse.end(),0.0)*scaleFactor);
+   }
+
+   //------------------------------------------------------------------------------------------------------------------
+   void CaloNoiseSimGenerator::initAR()
+   {
+       std::vector<double> wf(waveform_);
+       for (auto& val : wf) val -= pedestal_;
+
+       ARFitter_.setWaveform(wf);
+       ARFitter_.fitARCoeff();
+   }
+
+   //------------------------------------------------------------------------------------------------------------------
+   void CaloNoiseSimGenerator::generateFragments(const CaloWFExtractor& wfExtractor)
+   {
+       unsigned nwf(0), nfound(0), length(255);
+       for (;nwf<nMaxFragment_;++nwf)
+       {
+          std::vector<double> temp(length,0.0);
+          addFullNoise(temp,false);
+
+          std::vector<int> wf;
+          for (const auto& val : temp) wf.emplace_back(val - pedestal_);    
+
+          std::vector<unsigned> starts, stops;         
+          wfExtractor.extract(wf,starts,stops);
+          if (starts.empty()) continue;
+
+          std::vector<double> fragment;
+          std::copy(temp.begin()+starts[0], temp.begin()+stops[0]+1, std::back_inserter(fragment));
+          digiNoise_.push_back(fragment);
```

Comment:
Please check if std::copy is smart enough to figure out to do the reserve.  I bet it is but I don't know for sure.

Reply:
I don't believe it is (stl functions tend to do the minimum amount of work required). I have added a reserve. Note that this function was only called at the beginning of a subrun, and you would only copy a few short fragments, so the penalty in reallocating a few small vectors was small.

Reply:
Thanks.  My advise it just always do it, even beginJob, beginRun .... .  Then you don't forget when it's actually important.